cmd/debugmailbox/src/lib.rs

```diff
@@ -64,36 +68,82 @@ enum DebugMailboxCmd {
     Debug,
     /// Force the device into ISP mode
     Isp,
+    /// Get a Debug Authentication Challenge
+    AuthChallenge { out: PathBuf },
+    /// Send a Debug Authentication Response
+    AuthResponse { dar: PathBuf },
 }
 
-fn alive<'a>(
+fn poll_raw_ap_register<'a, F>(
     probe: &mut Box<dyn ArmProbeInterface + 'a>,
-    addr: &ApAddress,
-) -> Result<()> {
-    probe.write_raw_ap_register(*addr, CSW, 0x21)?;
+    ap: &ApAddress,
+    addr: u8,
+    f: F,
+    timeout_ms: usize,
+) -> Result<u32>
+where
+    F: Fn(u32) -> Result<bool>,
+{
+    let mut timeout_ms = timeout_ms;
 
-    let mut timeout = 100;
     loop {
+        // Sleep at least once to provide a little delay from any previous
+        // operation.  Really this should be done by the caller but it isn't and
+        // fixing all those cases is more annoying than having a little extra
+        // delay here.  Currently, read_return() will definitely fail Without
+        // this initial delay.
         std::thread::sleep(std::time::Duration::from_millis(10));
-        if let Ok(val) = probe.read_raw_ap_register(*addr, CSW) {
-            if val == 0 {
-                break;
+
+        match probe.read_raw_ap_register(*ap, addr) {
+            Ok(val) => match f(val) {
+                Ok(true) => return Ok(val),
+                Ok(false) => continue,
+                Err(e) => return Err(e),
+            },
+            Err(DebugProbeError::ArchitectureSpecific(arch_err)) => {
+                match arch_err.downcast::<DapError>() {
+                    Ok(dap_err) => match *dap_err {
+                        DapError::WaitResponse => {}
+                        e => {
+                            return Err(DebugProbeError::ArchitectureSpecific(
+                                e.into(),
+                            )
+                            .into())
+                        }
+                    },
+                    Err(e) => {
+                        return Err(
+                            DebugProbeError::ArchitectureSpecific(e).into()
+                        )
+                    }
+                }
             }
+            Err(x) => return Err(x.into()),
         }
 
-        if timeout == 0 {
+        timeout_ms = timeout_ms.saturating_sub(10);
+        if timeout_ms == 0 {
             break;
-        } else {
-            timeout -= -1;
         }
     }
 
-    if timeout == 0 {
-        bail!("Timed out waiting for reset. Chip is unlikely to be alive!");
-    } else {
-        println!("Reset chip successfully!");
-        Ok(())
+    Err(anyhow!("Timed out polling AP {ap:?} address 0x{addr:02x}"))
```

Comment:
(I notice this is your only use of `anyhow!` -- if you wanted to remove the import, this line is equivalent to `bail!`)